### ISSUE TITLE
SVG overlay using on("move") and project() out of sync in IE/Edge/Safari

### ISSUE THREAD

**Issue by joedjc** [mapbox/mapbox-gl-js#5390]
I'm trying to use an SVG overlay like this https://bl.ocks.org/shimizu/5f4cee0fddc7a64b55a9 or this http://bl.ocks.org/enjalot/0d87f32a1ccb9a720d29ba74142ba365.

I use a map.on("move") event to call a function which updates the location of the SVG elements, together with map.project() to get the on screen co-ordinates (similar to the first example - except without hiding the SVG graphics when moving).

In Chrome and Firefox this works well with the SVG graphics appearing seamlessly with mapbox, but in IE11, Edge and to some extent Safari there is a significant lag between the mouse move and the SVG features being updated.

Possibly this is related to this issue: https://github.com/mapbox/mapbox-gl-js/pull/4681 because I see similar behaviour with markers in those browsers. Not sure if that PR is being progressed?

**Example in Edge:**

![mapbox-svg](https://user-images.githubusercontent.com/13920398/31058500-998669cc-a6ec-11e7-86d3-6ef0f0e87304.gif)

**Comment by anandthakker** [comment]
Hi @joedjc, thanks for the report.  I believe this is a different issue from #4681.

It's interesting that, in the GIF above, it looks like it's the map lagging behind the SVG, not the other way around.  @joedjc how does the map's framerate look for you in Edge if you disable the `move` handler and just pan without the SVG layer being updated?

**Comment by joedjc** [comment]
Hi @anandthakker thanks for looking into it.

I think you might be onto something there...the SVG does indeed move before the map as though it fires the move event without checking that the map has caught up...or assumes the map can update at the same speed?

I'm sure you're right but the behaviour does look suspiciously similar with markers (e.g. https://github.com/mapbox/mapbox-gl-js/issues/5360). 

I tried testing a few things:

- In all browsers if I have a lot of SVG elements the frame rate slows (not surprising). However, in Chrome it still stays in sync (but at a lower frame rate) whereas the other browsers display the lag.
- It's noticeable in IE11/Edge even with just a small number of SVG features (which is where i'm having a problem). It's visible in Safari but less noticeable...I guess it corrects itself quite quickly.
- If I turn off the `move` hander, the performance is still quite slow in IE11/Edge with a lot of SVG shapes but is quick in other browsers.

Not sure what to conclude from that...but i'd be interested to know if the `move` handler is supposed to fire only once the map has actually moved?

**Comment by anandthakker** [comment]
We might have found the (or _an_) underlying issue here: the `DragPan` handler updates the camera directly and fires a `move` event.  Updates to the DOM that happen in `move` listeners happen synchronously, but the map renderer doesn't update until the next animation frame.

**Comment by dgrippisc** [comment]
oh nice.  this might be related to the issues we were seeing with `moveend` on mobile devices.  we just put in a workaround on friday:

```
...
    this.debouncedMoveEnd = debounce(
      () => { this.onMapMoveEnd(); },
      100,
    );

      // Mapboxgl doesn't not reliably fire `moveend` events.
      // This causes issues on mobile devices, where using multitouch zoom will consistently
      // fire `moveend` and crash Mobile Safari when using this event to update a url.
      //
      // Issue:
      //   https://github.com/mapbox/mapbox-gl-js/issues/3435
      //
      // Internal hack to fix this with the hash replaceState option:
      //   https://github.com/mapbox/mapbox-gl-js/pull/5613
      this.map.on('move', this.debouncedMoveEnd);
...
```

**Comment by joedjc** [comment]
@anandthakker Ok...I noticed this also affects zooming, rotating and tilting the map not just panning. Things like double clicking the map to zoom or pressing the compass to reset the map bearing also display this behaviour.

**Comment by anandthakker** [comment]
> Ok...I noticed this also affects zooming, rotating and tilting the map not just panning. Things like double clicking the map to zoom or pressing the compass to reset the map bearing also display this behaviour.

The zooming/rotating/panning are consistent with the above diagnosis, but not the double-click zoom or compass reset -- @joedjc any chance you could post a quick video / live example that exhibits this behavior in those cases?

**Comment by joedjc** [comment]
I've tried to capture it here - the red boxes are an svg overlay. I am actually using the zoom control buttons which effectively does the same animation as a double click. This is in Safari and running v0.42.2. It's generally more pronounced if the map is full screen. I actually see the lag when using the zoom controls in Chrome (which is generally ok in other cases).
![ezgif com-crop](https://user-images.githubusercontent.com/13920398/35056555-9ee400da-fbcb-11e7-969b-06385083df5c.gif)

**Comment by anandthakker** [comment]
#6005 attempts a fix for the drag pan/rotate/zoom cases of this behavior, and should be available in the next release (0.44).  

Let's keep this issue closed, as long as those cases are indeed fixed.  @joedjc if the zoom de-sync is still an issue on `master`, would you mind creating a new issue for it?

**Comment by jfirebaugh** [comment]
The underlying cause of the desynchronization here is that Safari and IE/Edge are non-spec compliant in running `requestAnimationFrame` callbacks _after_ the next recalc/layout/paint, rather than before. We'll keep our workaround in place, of course -- just wanted to make a note of that.

* https://bugs.webkit.org/show_bug.cgi?id=177484
* https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/15469349/

**Comment by Nagesh-Chandan** [comment]
@joedj:  _"I've tried to capture it here - the red boxes are an svg overlay"_ 
How are you loading the svg file using Source and Layer Or as an Annotation?
